plugins/base/src/main/kotlin/renderers/html/NavigationDataProvider.kt

```diff
@@ -0,0 +1,92 @@
+    open fun navigableChildren(input: RootPageNode): NavigationNode = input.withDescendants()
+        .first { it is ModulePage || it is MultimoduleRootPage }.let { visit(it as ContentPage) }
+
+    open fun visit(page: ContentPage): NavigationNode =
+        NavigationNode(
+            name = page.displayableName(),
+            dri = page.dri.first(),
+            sourceSets = page.sourceSets(),
+            icon = chooseNavigationIcon(page),
+            children = page.navigableChildren()
+        )
+
+    /**
+     * Parenthesis is applied in 1 case:
+     *  - page only contains functions (therefore documentable from this page is [DFunction])
+     */
+    private fun ContentPage.displayableName(): String =
+        if (this is WithDocumentables && documentables.all { it is DFunction }) {
+            "$name()"
+        } else {
+            name
+        }
+
+    private fun chooseNavigationIcon(contentPage: ContentPage): NavigationNodeIcon? {
+        return if (contentPage is WithDocumentables) {
+            val documentable = contentPage.documentables.firstOrNull()
+            val isJava = (documentable as? WithSources)?.sources?.any { it.value is PsiDocumentableSource } ?: false
+
+            when (documentable) {
+                is DClass -> when {
+                    documentable.isException -> NavigationNodeIcon.EXCEPTION
+                    documentable.isAbstract() -> {
+                        if (isJava) NavigationNodeIcon.ABSTRACT_CLASS else NavigationNodeIcon.ABSTRACT_CLASS_KT
+                    }
+                    else -> if (isJava) NavigationNodeIcon.CLASS else NavigationNodeIcon.CLASS_KT
+                }
+                is DFunction -> NavigationNodeIcon.FUNCTION
+                is DProperty -> {
+                    val isVar = documentable.extra[IsVar] != null
+                    if (isVar) NavigationNodeIcon.VAR else NavigationNodeIcon.VAL
+                }
+                is DInterface -> if (isJava) NavigationNodeIcon.INTERFACE else NavigationNodeIcon.INTERFACE_KT
+                is DEnum -> if (isJava) NavigationNodeIcon.ENUM_CLASS else NavigationNodeIcon.ENUM_CLASS_KT
+                is DAnnotation -> {
+                    if (isJava) NavigationNodeIcon.ANNOTATION_CLASS else NavigationNodeIcon.ANNOTATION_CLASS_KT
+                }
+                is DObject -> NavigationNodeIcon.OBJECT
+                else -> null
+            }
+        } else {
+            null
+        }
+    }
+
+    private fun DClass.isAbstract(): Boolean {
+        return modifier.values.all { it is KotlinModifier.Abstract || it is JavaModifier.Abstract }
+    }
+
+    private fun ContentPage.navigableChildren(): List<NavigationNode> {
+        return if (this !is ClasslikePageNode) {
+            children
+                .filterIsInstance<ContentPage>()
+                .map { visit(it) }
+                .sortedBy { it.name.toLowerCase() }
+        } else {
+            emptyList()
+        }
+    }
```

Comment:
Removed enum entries from here, so it's not in the navigation tree anymore

Reply:
Unfortunately, had to return enum entries to navigation. 

If someone has a direct link to an enum entry, it's unclear where to scroll - from entry's page I couldn't figure out a clean solution for selecting parent's navigation menu link. 

The only dirty way is to iterate all of the sideMenu and compare links, but that looks wrong on so many levels.

Maybe enum entries should not have pages at all. I'll bring this up during sync so we can brainstorm together